src/gromacs/nblib/topology.cpp

```diff
@@ -182,8 +298,41 @@ gmx::ListOfLists<int> TopologyBuilder::createExclusionsListOfLists() const
     return exclusionsListOfListsGlobal;
 }
 
+Topology::InteractionData TopologyBuilder::createInteractionData(const detail::ParticleSequencer& particleSequencer)
+{
+    auto create = [](auto& output, const auto& molecules, const auto& sequencer) {
+        using BondType = typename std::decay_t<decltype(output)>::type;
+        auto uniqueData = detail::eliminateDuplicateBonds(detail::aggregateBonds<BondType>(molecules));
```

Comment:
Instead of eliminating duplicates on the aggregated bond list, it seems like it would be simpler to just eliminate duplicates directly on the `molecules`. It seems like aggregate is only used here and that also removing duplicates would be simpler without aggregating first. So why have an aggregate?

Reply:
What you propose here is impossible. In order to recognize the duplicates, you must first sort the entire collection of bonds, if you want an N logN algorithm.

How would that even work to eliminate bonds directly in the molecule? How do you recognize the duplicates and keep track which ones were eliminated? If you start with the first molecule and compare against all the others and then move on to the second molecule, you'll be doing N^2 work. And then after you got rid of the duplicates, would you introduce cross-references to bonds between molecules?

It's still possible to improve the code; algorithmically, it's smarter to eliminate the bond duplicates in a list that is generated without expanding the bond list with multiple identical molecules. This will make the code a bit more complicated, but I'll see if it's possible to make this improvement without adding to much baggage.